lib/elixir/lib/calendar/datetime.ex

```diff
@@ -206,40 +319,167 @@ defmodule DateTime do
 
       iex> DateTime.from_naive!(~N[2016-05-24 13:26:08.003], "Etc/UTC")
       #DateTime<2016-05-24 13:26:08.003Z>
 
+      iex> DateTime.from_naive!(~N[2018-05-24 13:26:08.003], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      #DateTime<2018-05-24 13:26:08.003+02:00 CEST Europe/Copenhagen>
+
   """
   @doc since: "1.4.0"
-  @spec from_naive!(NaiveDateTime.t(), Calendar.time_zone()) :: t
-  def from_naive!(naive_datetime, time_zone) do
-    case from_naive(naive_datetime, time_zone) do
+  @spec from_naive!(
+          NaiveDateTime.t(),
+          Calendar.time_zone(),
+          Calendar.time_zone_database()
+        ) :: t
+  def from_naive!(
+        naive_datetime,
+        time_zone,
+        time_zone_database \\ Calendar.get_time_zone_database()
+      ) do
+    case from_naive(naive_datetime, time_zone, time_zone_database) do
       {:ok, datetime} ->
         datetime
 
+      {:ambiguous, dt1, dt2} ->
+        raise ArgumentError,
+              "cannot convert #{inspect(naive_datetime)} to datetime because such " <>
+                "instant is ambiguous in time zone #{time_zone} as there is an overlap " <>
+                "between #{inspect(dt1)} and #{inspect(dt2)}"
+
+      {:gap, dt1, dt2} ->
+        raise ArgumentError,
+              "cannot convert #{inspect(naive_datetime)} to datetime because such " <>
+                "instant does not exist in time zone #{time_zone} as there is a gap " <>
+                "between #{inspect(dt1)} and #{inspect(dt2)}"
+
       {:error, reason} ->
         raise ArgumentError,
-              "cannot parse #{inspect(naive_datetime)} to datetime, reason: #{inspect(reason)}"
+              "cannot convert #{inspect(naive_datetime)} to datetime, reason: #{inspect(reason)}"
+    end
+  end
+
+  @doc """
+  Changes the time zone of a `DateTime`.
+
+  Returns a `DateTime` for the same point in time, but instead at
+  the time zone provided. It assumes that `DateTime` is valid and
+  exists in the given timezone and calendar.
+
+  By default, it uses the default time zone database returned by
+  `Calendar.get_time_zone_database/1`, which defaults to
+  `Calendar.UTCOnlyTimeZoneDatabase` which only handles "Etc/UTC" datetimes.
+  Another time zone database can be passed as argument or set globally.
+  See the "Time zone database" section in the module docs.
+
+  ## Examples
+
+      iex> cph_datetime = DateTime.from_naive!(~N[2018-07-16 12:00:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> {:ok, pacific_datetime} = DateTime.shift_zone(cph_datetime, "America/Los_Angeles", FakeTimeZoneDatabase)
+      iex> pacific_datetime
+      #DateTime<2018-07-16 03:00:00-07:00 PDT America/Los_Angeles>
+
+  """
+  @doc since: "1.8.0"
+  @spec shift_zone(t, Calendar.time_zone(), Calendar.time_zone_database()) ::
+          {:ok, t} | {:error, :time_zone_not_found | :utc_only_time_zone_database}
+  def shift_zone(datetime, time_zone, time_zone_database \\ Calendar.get_time_zone_database())
```

Comment:
Should we have an accompanying `force_zone` function that uses the same naive dt, but in a specific zone? For example, if you got time in the wrong zone, but the numeric values are correct. From experience, I know such a function is useful.

If we don't want to provide such a function, maybe in the docs for `shift_zone` we describe how to do it by first converting to native and then from naive to the target zone.

Reply:
Sorry for not being clear. The use case I have in mind would be exactly equivalent to something like:

```elixir
def force_zone(datetime, time_zone) do
  from_naive(to_naive(datetime), time_zone)
end
```

This means basically that if I have a 9 AM CET and I want to have 9 AM UTC or similar. It's generally useful in many cases, where you actually don't want to do proper time zone conversion.

Reply:
@michalmuskala If you have a DateTime and want the same "wall time" in a different time zone you can pass it to a function only expecting a naive datetime. E.g. 2018-11-13 09:00:00 Europe/Copenhagen and want 9:00:00 for the same date but in UTC instead.

DateTime.from_naive(cph_dt_9oclock, "Etc/UTC") should return the same "wall time" as the one provided, but in UTC instead. Because `from_naive` just looks at the fields required by the Calendar.naive_datetime type and ignores the time zone part.